improver/utilities/mathematical_operations.py

```diff
@@ -200,66 +184,68 @@ def perform_integration(
-                    if upper_bound > self.end_point:
-                        continue
-                elif self.direction_of_integration == "negative":
-                    if lower_bound < self.end_point:
-                        continue
+        levels_tuple = zip(
+            upper_bounds_cube.slices_over(self.coord_name_to_integrate),
+            lower_bounds_cube.slices_over(self.coord_name_to_integrate))
+
+        for (upper_bounds_slice, lower_bounds_slice) in levels_tuple:
+            upper_bound, = upper_bounds_slice.coord(
+                self.coord_name_to_integrate).points
+            lower_bound, = lower_bounds_slice.coord(
+                self.coord_name_to_integrate).points
+
+            if skip_slice(upper_bound, lower_bound,
+                          self.direction_of_integration,
+                          self.start_point, self.end_point):
+                continue
+
             stride = np.abs(upper_bound - lower_bound)
-            upper_half_of_stride = upper_bounds_slice.data * 0.0
-            # Restrict the integration to only consider positive values.
-            # This condition is specific for the computation of the wet-bulb
-            # temperature integral.
-            uindex = np.where(upper_bounds_slice.data > 0)
-            upper_half_of_stride[uindex] = (upper_bounds_slice.data[uindex] *
-                                            0.5 * stride)
-            # Restrict the integration to only consider positive values.
-            # This condition is specific for the computation of the wet-bulb
-            # temperature integral.
-            lindex = np.where(lower_bounds_slice.data > 0)
-            lower_half_of_stride = lower_bounds_slice.data * 0.0
-            lower_half_of_stride[lindex] = (lower_bounds_slice.data[lindex] *
-                                            0.5 * stride)
-            stride_sum += lower_half_of_stride + upper_half_of_stride
-            integrated_slice.data = stride_sum
-            integrated_cubelist.append(integrated_slice.copy())
+            upper_half_data = np.where(
+                upper_bounds_slice.data > 0,
+                upper_bounds_slice.data * 0.5 * stride, 0.0)
+            lower_half_data = np.where(
+                lower_bounds_slice.data > 0,
+                lower_bounds_slice.data * 0.5 * stride, 0.0)
+            stride_sum += upper_half_data + lower_half_data
+
+            # Create new cube
+            template = (upper_bounds_slice
+                        if self.direction_of_integration == "positive"
+                        else lower_bounds_slice)
+            integrated_slice = create_new_diagnostic_cube(
+                template.name, template.units, template,
+                attributes, data=stride_sum.copy())
+            integrated_slice.coord(self.coord_name_to_integrate).bounds = (
+                np.array([lower_bound, upper_bound]).astype(coord_dtype))
+            integrated_cubelist.append(integrated_slice)
```

Comment:
Have you looked at what the computational cost of constructing a cube in every iteration is? We have previously discussed with the green team the desire to keep cubes away from maths processing where possible. I can see that this approach makes it nice and simple to construct the required coordinates, but I'd been interested in how efficient/inefficient it is.

Reply:
Will try doing this by creating arrays in the loop (data and height bounds) and putting them into the cube afterwards.